### PR TITLE
Hybrid Context Parallel Feature

### PR DESCRIPTION
**Pull request by parthmannan**
# What does this PR do ?
Dev MR for details - #2054 

Design document discussed in MCore sync meeting - https://docs.google.com/document/d/1MnIPQ_VbpDNp-adtvcEv-SYx6A8rtt3-fDdxbcdrmk0/edit?usp=sharing

The first issue this MR is trying to solve is the imbalance between DP ranks when using packed sequences (for example in SFT). While packing sequences can help reduce variability in total sequence length, it does not guarantee equal workload. Attention compute is quadratic to sequence length and a single long sequence of 1k has 2x more compute than a packed sequence made of 2x512 length. This problem gets much worse when we have very large sequences and/or a large variation between sequence lengths.
This MR schedules a variable number of microbatches per rank in DPxCP group to ensure balanced workload.

The second issue this MR is trying to solve is redundant CP communication. Our context parallel size is based on the full packed sequence length (usually the max seq length of all samples). For example, if a sequence of 1k requires CP2, we apply CP2 to a packed sequence of 2x512 as well. But in reality, we can easily partition the packed sequence of 2x512 into 2 GPUs by separating the 2 samples without any CP. This MR introduces dynamic context parallelism where each sample is individually scheduled with a dynamic CP group.

To achieve the above, we introduce a balanced scheduler and a dataloader wrapper.
The dataloader wrapper is responsible for collecting the metadata which informs the scheduler of the sequence length of each sample across the entire global batch. This dataloader breaks up the packed sequences into individual samples as they are individually scheduled. Once we have the metadata, we can perform the scheduling using the balanced scheduler which assigns samples to ranks (across DPxCP group) and a dynamic CP group size. To avoid any deadlocks, we divide the schedule into groups (this replaces the notion of microbatches). Within each group, each rank is part of a fixed CP group. However, each rank may run different number of samples in order for all ranks to have a balanced compute.

Screenshot 2025-10-08 at 3 21 39 PM

We have run performance and correctness evaluations using the feature. Using the SFT packed dataset with max seq len of 128k and testing with LLaMa3 8B dummy model, we see 3x performance improvement with this feature. While there is room for improving the baseline itself, the speedup should remain in the 2-3x range.

This is how 128k seq len with CP16 looks like (without this feature). The GPU is bound by CP communications.
Screenshot 2025-10-08 at 3 28 38 PM

This is how 128k seq len with CP16 looks like (with this feature). The GPU is bound by attention compute since all redundant comms have been removed.
Screenshot 2025-10-08 at 3 30 26 PM

Feature correctness (@xiaoyao0115)
hybrid_cp_loss_curve

This is the first milestone of this feature and there's many improvements that we want to make in the future releases.
1. The feature does not support pipeline parallelism or FSDP yet. We hope to add PP support next.
2. The feature is limited to creating dynamic groups of CP of power 2. We hope to add complete dynamic support using changes in TransformerEngine DPA.
3. The feature does not support CUDA graphs.
4. The feature works best with FlashAttention instead of cuDNN FusedAttention. This is because the changing lengths and CP size make cuDNN recompile the graph and all performance gains are lost. We'll advocate for dynamic support to cuDNN FusedAttention.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -2062,6 +2049,9 @@ def train(
     energy_monitor = get_energy_monitor()
     one_logger = get_one_logger()
 
+    if args.hybrid_context_parallel:
+        train_data_iterator = iter(HybridCPDataLoaderWrapper(train_data_iterator, config))
```
Comment (asolergi-nv) [inline megatron/training/training.py:2415]: Should we remove the `__iter__` method, WDYT?

```suggestion
        train_data_iterator = HybridCPDataLoaderWrapper(train_data_iterator, config)
```

File: megatron/core/models/common/embeddings/rotary_pos_embedding.py [hunk 2]
```diff
@@ -148,13 +148,12 @@ def get_cos_sin(self, max_seq_len: int, offset: int = 0) -> (Tensor, Tensor):
         return cos, sin
 
     @lru_cache(maxsize=32)
-    def forward(self, max_seq_len: int, offset: int = 0, packed_seq: bool = False) -> Tensor:
```
Comment (santhnm2) [inline megatron/core/models/common/embeddings/rotary_pos_embedding.py:151]: Will removing the `packed_seq` arg break backwards compatibility? If so can we keep it for now and add a deprecation warning? Same comment applies for the other rotary embedding classes.

File: megatron/core/extensions/transformer_engine.py [hunk 3]
```diff
@@ -1335,6 +1336,26 @@ def forward(
...
+            # If cp_group is None but local_cp_size is provided,
+            # Indicates to turn off CP dynamically
+            elif packed_seq_params.local_cp_size is not None:
+                assert (
+                    packed_seq_params.local_cp_size == 1
+                ), "local_cp_size must be == 1 if provided without cp_group"
+                super().set_context_parallel_group(None, None, None, self.cp_comm_type)
+            self.kept_packed_seq_params.discard("cp_group")
+            self.kept_packed_seq_params.discard("local_cp_size")
+
```
Comment (duncanriach) [inline megatron/core/extensions/transformer_engine.py:1357]: I assume there will not need to be changes to the Mamba packed sequence support (coming in [2913](https://github.com/NVIDIA/Megatron-LM/pull/2913)) to support this current feature.

File: megatron/core/models/common/embeddings/rotary_pos_embedding.py (hunk 2 above)
Comment (duncanriach) [reply megatron/core/models/common/embeddings/rotary_pos_embedding.py:151]: This will break `MambaModel`. As an alternative to keeping this here, could you change the call in `MambaModel`?

File: megatron/core/utils.py [hunk 4]
```diff
@@ -2055,6 +2073,102 @@ def get_batch_on_this_cp_rank(batch: Dict[str, Any]):
...
+        )
+        index = tex.thd_get_partitioned_indices(
+            cu_seqlens_padded, batch['tokens'].size(1), cp_size, cp_rank
+        )
+        for key, data in batch.items():
+            if key in {'attention_mask', 'cu_seqlens', 'cu_seqlens_padded', 'max_seqlen'}:
+                continue
+            batch[key] = data.index_select(1, index)
+
+    return batch, packed_seq_params
```
Comment (duncanriach) [inline megatron/core/utils.py:2113]: Noting that this code is a refactor of some code in pretrain_mamba.py in [2913](https://github.com/NVIDIA/Megatron-LM/pull/2913). Wondering about how those two things mesh.

File: megatron/training/utils.py
Comment (duncanriach) [inline megatron/training/utils.py:1]: There are a lot of changes in here that also overlap with [2913](https://github.com/NVIDIA/Megatron-LM/pull/2913). I think we should get [2913](https://github.com/NVIDIA/Megatron-LM/pull/2913) merged (which is very high priority) and then address any conflicts or issues with this PR

File: pretrain_gpt.py [hunk 5]
```diff
@@ -43,15 +43,29 @@ def get_batch(data_iterator, vp_stage=None):
...
+        # slice batch along sequence dimension for context parallelism
+        batch = get_batch_on_this_cp_rank(batch)  # The implementation of this function is in MCore
+        packed_seq_params = None
+    elif local_cp_size is None:  # Packed THD format
+        assert max_seqlen.dim() == 1
+        batch, packed_seq_params = get_thd_batch_on_this_cp_rank(batch, cu_seqlens, cu_seqlens_padded, max_seqlen)
+    else: # Hybrid CP format
+        batch, packed_seq_params = get_batch_on_this_hybrid_cp_rank(batch, local_cp_size)
+    
+    return (*batch.values(), packed_seq_params)
```
Comment (duncanriach) [inline pretrain_gpt.py:88]: The reason your changes here work with SFTDataset as the dataloader is that you've added the necessary changes to `pretrain_gpt.py`

**Review by duncanriach — COMMENTED** [review]
First pass review with a bunch of comments

File: megatron/training/utils.py
Comment (parthmannan) [reply megatron/training/utils.py:1]: Yes, we can merge 2913 first and then address conflicts here. One thing to keep in mind is that we are targeting 26.02 release for this feature which has a cut-off this Friday 1/16 so hopefully we can get 2913 asap to address conflicts in time.

File: megatron/core/utils.py (hunk 4 above)
Comment (parthmannan) [reply megatron/core/utils.py:2113]: I believe we should be able to use this function directly in `pretrain_mamba.py` like we do here in `pretrain_gpt.py`. But, since 2913 is going in first, we can do another PR to reduce code bloat in `pretrain_mamba.py` but it should not break anything.

File: megatron/core/extensions/transformer_engine.py (hunk 3 above)
Comment (parthmannan) [reply megatron/core/extensions/transformer_engine.py:1357]: If we want to support this feature for Mamba, we will need to update `pretrain_mamba.py` to make sure it has the right `packed_seq_params`. But nothing should break in Mamba and regular packed sequence can be used as currently in 2913.

File: megatron/core/models/common/embeddings/rotary_pos_embedding.py (hunk 2 above)
Comment (parthmannan) [reply megatron/core/models/common/embeddings/rotary_pos_embedding.py:151]: >Will removing the packed_seq arg break backwards compatibility? If so can we keep it for now and add a deprecation warning? Same comment applies for the other rotary embedding classes.

It will break backwards compatibility but the feature doesn't work with only a boolean packed_seq. It requires the dynamic CP group that's in the `packed_seq_params`. We could potentially keep `packed_seq` and add `cp_group` as an optional argument which would not break compatibility.
Let me explore that and test.

>This will break MambaModel. As an alternative to keeping this here, could you change the call in MambaModel?

MambaModel doesn't use the `packed_seq` as of now. After 2913, it would add that. So once 2913 is in, we can fix any change required in `MambaModel`

File: pretrain_gpt.py [hunk 6]
```diff
@@ -43,15 +43,29 @@ def get_batch(data_iterator, vp_stage=None):
     """Generate a batch."""
     # TODO: this is pretty hacky, find a better way
     if not is_first_or_last_pipeline_stage(vp_stage):
-        return None, None, None, None, None
+        return None, None, None, None, None, None
```
Comment (duncanriach) [inline pretrain_gpt.py:63]: Shouldn't this be 7 `None`s? One for `cu_seqlens` and one for `max_seqlen`?

Comment (duncanriach) [reply pretrain_gpt.py:63]: In [2913](https://github.com/NVIDIA/Megatron-LM/pull/2913), `get_batch` in `pretrain_mamba.py` returns `cu_seqlens` and `max_seqlen`. In this PR, `get_batch` in p`retrain_gpt.py` returns `packed_seq_params`. That's the difference. Not really an issue.

File: megatron/core/models/common/embeddings/rotary_pos_embedding.py (hunk 2 above)
Comment (parthmannan) [reply megatron/core/models/common/embeddings/rotary_pos_embedding.py:151]: Added the change to prevent breaking the API calls.

File: megatron/training/training.py (hunk 1 above)
Comment (parthmannan) [reply megatron/training/training.py:2415]: This is being removed in the part-2 PR (dev: #2000, main: #2304) where we make the dataloader interaction more flexible and modular.
I believe I required the extra `iter` here as some `train_data_iterator` were not iterable objects but just torch dataloader.

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/5fbdd32ea6831d4bc5b5b30700553c82224a5b69

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/e23a1e80042503097a822919a0e68f7f362e4da0

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/465be7ca9f3a53ee359ecef6810143c341a33295

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/491702cf7c1bb6943cd0449f3bf4e3400ae1257e

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/1860dbeb5d77bb05ad6926f4a897c0b6be6c2a06

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/4a1723f47d4867b92c8b409ad58f3b49f54c2c91

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/06d086fe684b2ef4243ed04bb5f3da744666e9f4

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/b09fb1bba9c1a4a8ab298b3d0610b6a8983d618f

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/1ebcb02dcb1043786572223eb54b152f60f26760

**Comment by parthmannan** [comment]
Functional test suite has been run and passing on Gitlab in https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/pipelines/41910205

**Comment by parthmannan** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2282/commits/641b04db3c6a4d6d274a4dfdad1a36e1aefc0ced
